src/server.rs

```diff
@@ -0,0 +1,294 @@
+use std::net::SocketAddr;
+use std::sync::Arc;
+
+use tokio::net::UdpSocket;
+use tokio::sync::mpsc::{self, UnboundedReceiver, UnboundedSender};
+use tokio::sync::Mutex;
+use tokio::task::JoinHandle;
+
+use crate::ScuttleButt;
+
+/// Buffer size for UDP messages.
+/// This value is set to avoid UDP fragmentation.
+/// https://jvns.ca/blog/2017/02/07/mtu/
+const BUF_SIZE: usize = 1_472;
+
+/// UDP ScuttleButt server.
+pub struct ScuttleServer {
+    channel: UnboundedSender<ChannelMessage>,
+    scuttlebutt: Arc<Mutex<ScuttleButt>>,
+    join_handle: JoinHandle<Result<(), anyhow::Error>>,
+}
+
+impl ScuttleServer {
+    /// Launch a new server.
+    ///
+    /// This will start the ScuttleButt server as a new Tokio background task.
+    pub fn spawn(address: impl Into<String>, seed_nodes: Vec<String>) -> Self {
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        let scuttlebutt = Arc::new(Mutex::new(ScuttleButt::with_node_id(address.into())));
+        let scuttlebutt_server = scuttlebutt.clone();
+
+        let join_handle = tokio::spawn(async move {
+            let mut server = UdpServer::new(rx, scuttlebutt_server).await?;
+            server.run().await
+        });
+
+        let scuttle_server = Self {
+            channel: tx,
+            scuttlebutt,
+            join_handle,
+        };
+
+        // Advertise the new node to all seed nodes.
+        for node_id in seed_nodes {
+            let _ = scuttle_server.gossip(node_id);
+        }
+
+        scuttle_server
+    }
+
+    /// Call a function with access to the [`ScuttleButt`].
+    pub async fn with_scuttlebutt<F, T>(&self, mut fun: F) -> T
+    where
+        F: FnMut(&ScuttleButt) -> T,
+    {
+        let scuttlebutt = self.scuttlebutt.lock().await;
+        fun(&scuttlebutt)
+    }
+
+    /// Call a function with mutable access to the [`ScuttleButt`].
+    ///
+    /// This will automatically gossip about the changes to all known nodes.
+    pub async fn with_scuttlebutt_mut<F, T>(&self, mut fun: F) -> T
+    where
+        F: FnMut(&mut ScuttleButt) -> T,
+    {
+        let mut scuttlebutt = self.scuttlebutt.lock().await;
+        let retval = fun(&mut scuttlebutt);
+
+        // Broadcast changes to all nodes after mutation.
+        for node in scuttlebutt.cluster_state_map.nodes() {
+            let _ = self.gossip(node);
+        }
+
+        retval
+    }
+
+    /// Shut the server down.
+    pub async fn shutdown(self) -> Result<(), anyhow::Error> {
+        let _ = self.channel.send(ChannelMessage::Shutdown);
+        self.join_handle.await?
+    }
+
+    /// Perform a ScuttleButt "handshake" with another UDP server.
+    pub fn gossip(&self, addr: impl Into<String>) -> Result<(), anyhow::Error> {
+        Ok(self.channel.send(ChannelMessage::Gossip(addr.into()))?)
```

Comment:
```suggestion
    pub fn gossip(&self, addr: impl Into<String>) -> anyhow::Result<()> {
        self.channel.send(ChannelMessage::Gossip(addr.into()))?;
        Ok(())
```

`.context()` could be nice too.

Reply:
I'm not familiar with `anyhow` or its `context` method, but I feel like any handling of this error should make the source quite obvious? If you have specific information you'd like to attach in the context I'm open to adding it though.

Reply:
ok